src/consensus/blockchain.py

```diff
@@ -102,42 +105,47 @@ async def _load_chain_from_store(self) -> None:
         """
         Initializes the state of the Blockchain class from the database.
         """
-        self.sub_blocks, peak = await self.block_store.get_sub_block_records()
-        self.sub_height_to_hash = {}
-        self.sub_epoch_summaries = {}
+        self.__sub_blocks, peak = await self.block_store.get_sub_block_records()
+        self.__sub_heights_in_cache = {}
+        self.__sub_height_to_hash = {}
+        self.__sub_epoch_summaries = {}
 
-        if len(self.sub_blocks) == 0:
+        if len(self.__sub_blocks) == 0:
             assert peak is None
             self.peak_height = None
             return
 
         assert peak is not None
-        self.peak_height = self.sub_blocks[peak].sub_block_height
+        self.peak_height = self.__sub_blocks[peak].sub_block_height
 
         # Sets the other state variables (peak_height and height_to_hash)
-        curr: SubBlockRecord = self.sub_blocks[peak]
+        curr: SubBlockRecord = self.__sub_blocks[peak]
         while True:
-            self.sub_height_to_hash[curr.sub_block_height] = curr.header_hash
+            self.__sub_height_to_hash[curr.sub_block_height] = curr.header_hash
             if curr.sub_epoch_summary_included is not None:
-                self.sub_epoch_summaries[curr.sub_block_height] = curr.sub_epoch_summary_included
+                self.__sub_epoch_summaries[curr.sub_block_height] = curr.sub_epoch_summary_included
             if curr.sub_block_height == 0:
                 break
-            curr = self.sub_blocks[curr.prev_hash]
-        assert len(self.sub_height_to_hash) == self.peak_height + 1
+            #  only keep last SUB_BLOCKS_CACHE_SIZE in mem
+            if len(self.__sub_blocks) < self.constants.SUB_BLOCKS_CACHE_SIZE:
+                curr = self.__sub_blocks[curr.prev_hash]
+            self.__sub_heights_in_cache[curr.sub_block_height] = []
+            self.__sub_heights_in_cache[curr.sub_block_height].append(curr.header_hash)
```

Comment:
You can do:
            self.__sub_heights_in_cache[curr.sub_block_height] = [curr.header_hash]